base/loading.jl

```diff
@@ -2555,7 +2562,21 @@ end
 root_module_exists(key::PkgId) = @lock require_lock haskey(loaded_modules, key)
 maybe_root_module(key::PkgId) = @lock require_lock get(loaded_modules, key, nothing)
 
-loaded_modules_array() = @lock require_lock copy(loaded_modules_order)
+function loaded_modules_array()
+    @lock require_lock begin
+        ret = copy(loaded_modules_order)
+        if generating_output(false)
+            # If we're generating a non-incremental system image, also add the
+            # Compiler to the module list so that we generate native code for it.
+            # However, we do not add it to `loaded_modules_order` unless there
+            # is a Julia-level dependency on it. The (non-)loadedness of the compiler
+            # is an implementation detail and we do not want ordinary pkgimages to
+            # acquire a dependency on the Compiler if they do not use it internally.
+            push!(ret, Compiler)
+        end
+        return ret
+    end
+end
```

Comment:
Returning the wrong answer most of the time to this query can cause inference to generate incorrect answers unpredictably, so let's not lie to the compiler. If you don't want this method to show up in InteractiveUtils, then maybe just filter it from there?

Each toplevel module must show up exactly once in this list, or things will go badly.